### PR TITLE
Implement sceNpSignaling & signaling improvements

### PR DESCRIPTION
**Pull request by RipleyTom**
If you want to test:
- RPCN Test server is test-np.rpcs3.net
- You need to make an account(email isn't checked, you can just click Create Account after changing host to test server)
- This server is temporary and can be taken down at any time
- Don't use the test server to connect to DeS as you won't be able to connect to other users.

What's in the PR:
- Improved signaling
Now for games that use sceNpMatching2Signaling connectivity is ensured by address negociation if negociation is possible, it should improve connectivity between nat1/2 and nat3 users(though ofc nat3/nat3 is still impossible).
- Implements sceNpSignaling
Some games use an older version of signaling and it's now implemented.
- Fixes, so many fixes!
STREAM_P2P sockets are now properly implemented, fixed many bugs all over the place.
Special thanks to dio who helped me squash those bugs through many, many hours of testing.

What games now work?
In addition to DeS and Bomberman, Scott Pilgrim multiplayer now works.

### CODE REVIEW

File: rpcs3/Emu/NP/np_handler.cpp [hunk 1]
```diff
@@ -556,7 +573,7 @@ void np_handler::operator()()
 			case NotificationType::UserJoinedRoom: notif_user_joined_room(notif.second); break;
 			case NotificationType::UserLeftRoom: notif_user_left_room(notif.second); break;
 			case NotificationType::RoomDestroyed: notif_room_destroyed(notif.second); break;
-			case NotificationType::SignalP2PEstablished: notif_p2p_established(notif.second); break;
+			case NotificationType::SignalP2PConnect: notif_p2p_connect(notif.second); break;
 			case NotificationType::RoomMessageReceived: notif_room_message_received(notif.second); break;
```
Comment (bevanweiss) [inline rpcs3/Emu/NP/np_handler.cpp:577]: Is this missing the '_SignalP2PDisconnect' notification?
Even if it is just an empty break..

Comment (RipleyTom) [reply rpcs3/Emu/NP/np_handler.cpp:577]: It's not implemented on server so no point atm(_ is a prefix for unused items in rust).

File: rpcs3/Emu/Cell/lv2/sys_net.cpp [hunk 2]
```diff
@@ -1010,23 +1035,41 @@ s32 send_packet_from_p2p_port(const std::vector<u8>& data, const sockaddr_in& ad
...
+std::vector<std::pair<std::pair<u32, u16>, std::vector<u8>>> get_sign_msgs()
+{
+	auto msgs     = std::vector<std::pair<std::pair<u32, u16>, std::vector<u8>>>();
+	const auto nc = g_fxo->get<network_context>();
+	{
+		std::lock_guard list_lock(nc->list_p2p_ports_mutex);
+		auto& def_port = nc->list_p2p_ports.at(3658);
+		{
+			std::lock_guard lock(def_port.s_sign_mutex);
+			msgs               = std::move(def_port.sign_msgs);
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_net.cpp:1064]: Lol that gap

File: rpcs3/Emu/Cell/lv2/sys_net.cpp [hunk 3]
```diff
@@ -2089,29 +2137,32 @@ error_code sys_net_bnet_recvfrom(ppu_thread& ppu, s32 s, vm::ptr<void> buf, u32
...
 				}
 
-				native_result = std::min(sock.p2ps.data_available, len);
-				memcpy(buf.get_ptr(), sock.p2ps.received_data.data(), native_result);
-
-				sock.p2ps.data_beg_seq += native_result;
-				for(auto it = sock.p2ps.data_mapping.begin(); it != sock.p2ps.data_mapping.end();)
+				const u32 to_give = std::min(sock.p2ps.data_available, len);
+				sys_net.trace("STREAM-P2P socket had %d available, given %d", sock.p2ps.data_available, to_give);
+
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_net.cpp:2142]: Spacing

File: rpcs3/Emu/NP/rpcn_client.h [hunk 4]
```diff
@@ -120,6 +120,7 @@ enum CommandType : u16
 	SetRoomDataInternal,
 	PingRoomOwner,
 	SendRoomMessage,
+    RequestSignalingInfos,
```
Comment (Megamouse) [inline rpcs3/Emu/NP/rpcn_client.h:123]: Indentation

File: rpcs3/rpcs3qt/rpcn_settings_dialog.cpp [hunk 5]
```diff
@@ -37,6 +37,7 @@ rpcn_settings_dialog::rpcn_settings_dialog(QWidget* parent)
 	QPushButton* btn_chg_pass = new QPushButton(tr("Set Password"));
 	QLabel *label_token       = new QLabel(tr("Token:"));
 	m_edit_token              = new QLineEdit();
+	m_edit_token->setMaxLength(16);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:40]: Is this magic number not available as a constant somewhere in rpcn code?

Comment (RipleyTom) [reply rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:40]: No, it's something completely unrelated to ps3, it's token to verify email...

File: rpcs3/Emu/Cell/Modules/sceNp.cpp [hunk 6]
```diff
@@ -4483,6 +4527,13 @@ error_code sceNpSignalingGetConnectionStatus(u32 ctx_id, u32 conn_id, vm::ptr<s3
 		return SCE_NP_SIGNALING_ERROR_INVALID_ARGUMENT;
 	}
 
+	const auto sigh = g_fxo->get<named_thread<signaling_handler>>();
+	const auto si = sigh->get_sig_infos(conn_id);
+
+	*conn_status           = si.connStatus;
+	(*peer_addr).np_s_addr = si.addr; // infos.addr is already BE
+	*peer_port             = si.port;
```
Comment (bevanweiss) [inline rpcs3/Emu/Cell/Modules/sceNp.cpp:4535]: peer_addr and peer_port deference without test?

File: rpcs3/Emu/Cell/Modules/sceNp2.cpp [hunk 7]
```diff
@@ -423,11 +423,12 @@ error_code sceNpMatching2SignalingGetConnectionStatus(
...
-	const auto& infos = nph->get_peer_infos(ctxId, roomId, memberId);
+	const auto sigh = g_fxo->get<named_thread<signaling_handler>>();
+	const auto si = sigh->get_sig2_infos(roomId, memberId);
 
-	*connStatus           = infos.connStatus;
-	(*peerAddr).np_s_addr = infos.addr; // infos.addr is already BE
-	*peerPort             = std::bit_cast<u16, be_t<u16>>(infos.port); // infos.port is already BE
+	*connStatus           = si.connStatus;
+	(*peerAddr).np_s_addr = si.addr; // infos.addr is already BE
+	*peerPort             = si.port;
```
Comment (bevanweiss) [inline rpcs3/Emu/Cell/Modules/sceNp2.cpp:431]: connStatus, peerAddr, peerPort deference without test? (or is my github view bust)

**Pull request merged**